File: javascript/Go/javascript_base_parser.go
package parser

import (
	"strings"

	"github.com/antlr/antlr4/runtime/Go/antlr"
)

// JavaScriptBaseParser implementation.
type JavaScriptBaseParser struct {
	*antlr.BaseParser
}

// Short for p.prev(str string)
func (p *JavaScriptBaseParser) p(str string) bool {
	return p.prev(str)
}

// Whether the previous token value equals to str.
func (p *JavaScriptBaseParser) prev(str string) bool {
	return p.GetTokenStream().LT(-1).GetText() == str
}

// Short for p.next(str string)
func (p *JavaScriptBaseParser) n(str string) bool {
	return p.next(str)
}

// Whether the next token value equals to str.
func (p *JavaScriptBaseParser) next(str string) bool {
	return p.GetTokenStream().LT(1).GetText() == str
}

func (p *JavaScriptBaseParser) notLineTerminator() bool {
	return !p.here(JavaScriptParserLineTerminator)
}

func (p *JavaScriptBaseParser) notOpenBraceAndNotFunction() bool {
	nextTokenType := p.GetTokenStream().LT(1).GetTokenType()
	return nextTokenType != JavaScriptParserOpenBrace && nextTokenType != JavaScriptParserFunction
}

func (p *JavaScriptBaseParser) closeBrace() bool {
	return p.GetTokenStream().LT(1).GetTokenType() == JavaScriptParserCloseBrace
}

// Returns true if on the current index of the parser's
// token stream a token of the given type exists on the
// Hidden channel.
func (p *JavaScriptBaseParser) here(_type int) bool {
	// Get the token ahead of the current index.
	possibleIndexEosToken := p.GetCurrentToken().GetTokenIndex() - 1
	ahead := p.GetTokenStream().Get(possibleIndexEosToken)

	// Check if the token resides on the HIDDEN channel and if it's of the
	// provided type.
	return ahead.GetChannel() == antlr.LexerHidden && ahead.GetTokenType() == _type
}

// Returns true if on the current index of the parser's
// token stream a token exists on the Hidden channel which
// either is a line terminator, or is a multi line comment that
// contains a line terminator.
func (p *JavaScriptBaseParser) lineTerminatorAhead() bool {
	// Get the token ahead of the current index.
	possibleIndexEosToken := p.GetCurrentToken().GetTokenIndex() - 1
	ahead := p.GetTokenStream().Get(possibleIndexEosToken)

	if ahead.GetChannel() != antlr.LexerHidden {
		// We're only interested in tokens on the HIDDEN channel.
		return true
	}

	if ahead.GetTokenType() == JavaScriptParserLineTerminator {
		// There is definitely a line terminator ahead.
		return true
	}

	if ahead.GetTokenType() == JavaScriptParserWhiteSpaces {
		// Get the token ahead of the current whitespaces.
		possibleIndexEosToken = p.GetCurrentToken().GetTokenIndex() - 2
		ahead = p.GetTokenStream().Get(possibleIndexEosToken)
	}

	// Get the token's text and type.
	text := ahead.GetText()
	_type := ahead.GetTokenType()

	// Check if the token is, or contains a line terminator.
	return (_type == JavaScriptParserMultiLineComment && (strings.Contains(text, "\r") || strings.Contains(text, "\n"))) ||
		(_type == JavaScriptParserLineTerminator)
}

File: javascript/JavaScript/JavaScriptBaseParser.js
const antlr4 = require("antlr4/index");
const JavaScriptParser = require("./JavaScriptParser");

function JavaScriptBaseParser(input) {
    antlr4.Parser.call(this, input);
}

JavaScriptBaseParser.prototype = Object.create(antlr4.Parser.prototype);

JavaScriptBaseParser.prototype.p = function(str) {
    return this.prev(str);
};

JavaScriptBaseParser.prototype.prev = function(str) {
    const source = this._input.LT(-1).source[1].strdata;
    const start = this._input.LT(-1).start;
    const stop = this._input.LT(-1).stop;
    const prev = source.slice(start, stop + 1);
    return prev === str;
};

JavaScriptBaseParser.prototype.notLineTerminator = function() {
    return !this.here(JavaScriptParser.LineTerminator);
};

JavaScriptBaseParser.prototype.notOpenBraceAndNotFunction = function() {
    const nextTokenType = this._input.LT(1).type;
    return (
        nextTokenType !== JavaScriptParser.OpenBrace &&
        nextTokenType !== JavaScriptParser.Function
    );
};

JavaScriptBaseParser.prototype.closeBrace = function() {
    return this._input.LT(1).type === JavaScriptParser.CloseBrace;
};

JavaScriptBaseParser.prototype.here = function(type) {
    const possibleIndexEosToken = this.getCurrentToken().tokenIndex - 1;
    const ahead = this._input.get(possibleIndexEosToken);
    return ahead.channel === antlr4.Lexer.HIDDEN && ahead.type === type;
};

JavaScriptBaseParser.prototype.lineTerminatorAhead = function() {
    let possibleIndexEosToken = this.getCurrentToken().tokenIndex - 1;
    let ahead = this._input.get(possibleIndexEosToken);
    if (ahead.channel !== antlr4.Lexer.HIDDEN) {
        return false;
    }

    if (ahead.type === JavaScriptParser.LineTerminator) {
        return true;
    }

    if (ahead.type === JavaScriptParser.WhiteSpaces) {
        possibleIndexEosToken = this.getCurrentToken().getTokenIndex() - 2;
        ahead = this._input.get(possibleIndexEosToken);
    }

    const text = ahead.type;
    const type = ahead.type;

    return (
        (type === JavaScriptParser.MultiLineComment &&
            (text.includes("\r") || text.includes("\n"))) ||
        type === JavaScriptParser.LineTerminator
    );
};

module.exports.JavaScriptBaseParser = JavaScriptBaseParser;
